### PR TITLE
Implement url support for RouteInformation and didPushRouteInformation

### PR DESCRIPTION
**Pull request by chunhtai**
Related https://github.com/flutter/flutter/issues/100624

The goal is to make sure the engine can send a location string in either the existing format or a complete uri string to the framework, and the framework will still work as usual.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/router.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/router.dart:47]:
```suggestion
      'Pass Uri.parse(location) to uri parameter instead. '
```

File: packages/flutter/lib/src/widgets/router.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/router.dart:48]: Update this?

File: packages/flutter/lib/src/widgets/router.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/router.dart:64]: Should this getter be deprecated as well? Or do you plan to keep it around even after the constructor arg is removed?

File: packages/flutter/lib/src/widgets/router.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/router.dart:80]: Does this still make sense now that the type is Uri and not String?

File: packages/flutter/lib/src/widgets/binding.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/binding.dart:650]: What is the purpose of this PR? It seems like all it effectively does is parse the string here into an URI, only to convert it back to a string above in line 91. It never does anything useful with the URI representation?

(Or is this just meant as a check that `location` has proper URI format? In that case, should we just do the parsing in an assert? Or is this user-provided input and we need to always verify this even in release mode?)

File: packages/flutter/lib/src/widgets/router.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/router.dart:44]: This is no longer true, either uri or location must be non-null.

File: packages/flutter/lib/src/widgets/binding.dart
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/binding.dart:650]: The purpose is to prepare this api so that engine can start sending full uri like `http://www.mywebsite.com/mypath`. This is important for apps that want to handle deeplinks from different web domains. Right now if we do it directly this will be a breaking change.

The line 91 is for backward compatibility that does not use router API.

**Comment by chunhtai** [comment]
The linux flutter_plugins ci is blocked on https://github.com/flutter/plugins/pull/7133, please ignore

File: packages/flutter/lib/src/widgets/router.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/router.dart:82]: same question as above.

File: packages/flutter/lib/src/widgets/binding.dart
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/binding.dart:99]: The goal is that Android and iOS can start sending
`http://mydomain.com/mypath` instead of just `/mypath` when it is deeplinked without breaking the existing code.

File: packages/flutter/lib/src/widgets/binding.dart
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/binding.dart:93]: Uri.parse will encode the string, so if you do `http://host.com/my path` The path will be `/my%20path` which will break the existing code

File: packages/flutter/lib/src/widgets/binding.dart
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/binding.dart:659]: This pr also add new getter `RouteInformation.uri`.

It lets developer to query RouteInformation.uri.scheme or RouteInformation.uri.host when they write their routing code.

**Comment by goderbauer** [comment]
This seems just hack-y. On the one hand, it tries to pretend that everything is an URI now. On the other hand, it very much still relies on the fact that sometimes it isn't (`_getLocationFromUri`). This is just weird and seems super-hard to properly reason about.

**Comment by chunhtai** [comment]
Let's define `location` = _getLocationFromUri(uri);

Only didpushRoute and RouteInformation.location use `location`, and `RouteInformation.location` is about to be deprecated, so basically only didpushRoute uses it. everything else should be uri otherwise.

The engine may send both `location` and `uri`, but they would be parse as if they are both uri, and I think `location` is just an `uri` without scheme and host.

What's other alternative? We can introduce a new RouteUri class and have a set of new method for didPushRouteUri, but I think that is redundent since RouteInformation is a class and can be extends to carry more data already.

**Comment by jjoelson** [comment]
Hi @chunhtai! I have a use-case in a mobile app where we're using a `NavigatorObserver` to report analytics whenever navigation changes, and the analytics team would like us to report if a particular navigation change was triggered by a deep link and if so then what was the full URL of the deep link. Currently we have to use a plugin like `uni_links` to gain access to the full URL. Do you think this change would enable us to access the full URL in our `RouteInformationParser` and switch to Flutter's built-in deep link handling?

**Comment by chunhtai** [comment]
@jjoelson yes that should be supported once the github issue is supported

**Comment by chunhtai** [comment]
will write a migration guide once this is approved

File: packages/flutter/lib/src/widgets/binding.dart
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/binding.dart:708]: Fix this?

File: packages/flutter/lib/src/widgets/app.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/app.dart:1500]: Why can we just ignore all other pieces of the URI here? Should we assert that those are empty?

File: packages/flutter/lib/src/widgets/binding.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/binding.dart:694]: Parse this once outside of the for loop?

**Review by goderbauer — APPROVED** [review]
LGTM

**Comment by chunhtai** [comment]
would need to migrate some internal code and go_router in the internal codebase is outdated as well

**Comment by chunhtai** [comment]
devtool also need to be rolled cl/525207188

**Comment by chunhtai** [comment]
g3fix cl/525552584

**Pull request merged**
